includes/Core/Authentication/Authentication.php

```diff
@@ -405,6 +405,25 @@ function ( $user_id ) {
 			add_action( 'googlesitekit_authorize_user', $set_initial_version );
 			add_action( 'googlesitekit_reauthorize_user', $set_initial_version );
 		}
+
+		$maybe_refresh_token = function() {
+			$token = $this->token->get();
+
+			// Do nothing if the token is not set.
+			if ( empty( $token['created'] ) || empty( $token['expires_in'] ) ) {
+				return;
+			}
+
+			// Do nothing if the token expires in more than 5 minutes.
+			if ( $token['created'] + $token['expires_in'] > time() + 5 * MINUTE_IN_SECONDS ) {
+				return;
+			}
+
+			$this->get_oauth_client()->refresh_token();
+		};
+
+		add_action( 'admin_init', $maybe_refresh_token );
+		add_action( 'heartbeat_tick', $maybe_refresh_token );
```

Comment:
It would be _kind of_ nicer to only do this when in any Site Kit-related UI (e.g. Site Kit screen, WP dashboard, etc.). But that may require a bunch of overhead in engineering - and realistically, most people land in the WP dashboard after logging in anyway, which should be the most common scenario where this logic would trigger one way or another.